base/special/trig.jl

```diff
@@ -22,65 +22,328 @@ end
     S1 = -0.16666666641626524
     S2 = 0.008333329385889463
-    S3 = -0.00019839334836096632
-    S4 = 2.718311493989822e-6
-
-    z = x.hi*x.hi
+    z = y.hi*y.hi
     w = z*z
-    r = S3+z*S4
-    s = z*x.hi
-    Float32((x.hi + s*(S1+z*S2)) + s*w*r)
+    r = @horner(z, -0.00019839334836096632, 2.718311493989822e-6)
+    s = z*y.hi
+    Float32((y.hi + s*@horner(z, S1, S2)) + s*w*r)
 end
 
-function cos_kernel(x::DoubleFloat32)
+# cos methods
+@noinline cos_domain_error(x) = throw(DomainError(x, "cos(x) is only defined for finite x."))
+@inline function cos(x::T) where T<:Union{Float32, Float64}
+    absx = abs(x)
+    if absx < T(pi)/4
+        if absx < sqrt(eps(T)/T(2.0))
+            return T(1.0)
+        end
+        return cos_kernel(x)
+    elseif isnan(x)
+        return T(NaN)
+    elseif isinf(x)
+        cos_domain_error(x)
+    else
+        n, y = rem_pio2_kernel(x)
+        n = n&3
+        if n == 0
+            return cos_kernel(y)
+        elseif n == 1
+            return -sin_kernel(y)
+        elseif n == 2
+            return -cos_kernel(y)
+        else
+            return sin_kernel(y)
+        end
+    end
+end
+
+const DC1 = 4.16666666666666019037e-02
+const DC2 = -1.38888888888741095749e-03
+const DC3 = 2.48015872894767294178e-05
+const DC4 = -2.75573143513906633035e-07
+const DC5 = 2.08757232129817482790e-09
+const DC6 = -1.13596475577881948265e-11
+"""
+    cos_kernel(y)
+
+Compute the cosine on the interval y∈[-π/4; π/4].
+"""
+@inline function cos_kernel(y::DoubleFloat64)
+    y² = y.hi*y.hi
+    y⁴ = y²*y²
+    r  = y²*@horner(y², DC1, DC2, DC3) + y⁴*y⁴*@horner(y², DC4, DC5, DC6)
+    half_y² = 0.5*y²
+    w  = 1.0-half_y²
+    w + (((1.0-w)-half_y²) + (y²*r-y.hi*y.lo))
+end
+@inline function cos_kernel(y::Float64)
+    y² = y*y
+    y⁴ = y²*y²
+    r  = y²*@horner(y², DC1, DC2, DC3) + y⁴*y⁴*@horner(y², DC4, DC5, DC6)
+    half_y² = 0.5*y²
+    w  = 1.0-half_y²
+    w + (((1.0-w)-half_y²) + (y²*r))
+end
+
+# cos_kernels accepting values from rem_pio2 in the Float32 case
+cos_kernel(x::Float32) = cos_kernel(DoubleFloat32(x))
+@inline function cos_kernel(y::DoubleFloat32)
     C0 = -0.499999997251031
     C1 = 0.04166662332373906
-    C2 = -0.001388676377460993
-    C3 = 2.439044879627741e-5
+    y² = y.hi*y.hi
+    y⁴ = y²*y²
+    r = @horner(y², -0.001388676377460993, 2.439044879627741e-5)
+    Float32(((1.0+y²*C0) + y⁴*C1) + (y⁴*y²)*r)
+end
 
-    z = x.hi*x.hi
-    w = z*z
-    r = C2+z*C3
-    Float32(((1.0+z*C0) + w*C1) + (w*z)*r)
+### sincos methods
+@noinline sincos_domain_error(x) = throw(DomainError(x, "sincos(x) is only defined for finite x."))
+
+"""
+    sincos(x)
+
+Simultaneously compute the sine and cosine of `x`, where the `x` is in radians.
+"""
+@inline function sincos(x::T) where T<:Union{Float32, Float64}
+    if abs(x) < T(pi)/4
+        if x == zero(T)
+            return x, one(T)
+        end
+        return sincos_kernel(x)
+    elseif isnan(x)
+        return T(NaN)
```

Comment:
This is the issue, should return `T(NaN), T(NaN)`.

Reply:
Thanks for catching this!

Reply:
You're welcome, you made an impressive work!  Is this a faithful translation from openlibm or are there other improvements?  The speedup is awesome.